### PR TITLE
avoid `Any32` type in `filter` for tuples

### LINKED ISSUE #42236: compiler hang in julia 1.7 on Gridap tutorial
See the following issue in Gridap.jl — one of the tutorials that worked in Julia 1.6 now causes the compiler to hang in 1.7-beta4

https://github.com/gridap/Tutorials/issues/118

cc @fverdugo

### PR DESCRIPTION
**Pull request by JeffBezanson**
This is sufficient to fix the case in that issue, but we should probably stop using these types. After intersection, they tend to blow up into more complex long tuple types (tuples of many unions) that are intractable for subtyping. With some quick experiments, so far it looks to me like we can replace these with length checks now with no regressions.

fixes #42236

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]

**Review by timholy — APPROVED** [review]

**Review by ararslan — APPROVED** [review]

**Review by Sacha0 — APPROVED** [review]
Nice! :)

**Comment by ararslan** [comment]
I think this should have had at least several more approvals before merging, 4 is not enough

**Pull request merged**
